sound/soc/sof/ipc4-topology.c

```diff
@@ -1543,6 +1543,12 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 		out_ref_channels = params_channels(fe_params);
 		break;
 	default:
+		out_ref_valid_bits = sof_ipc4_get_valid_bits(sdev, fe_params);
+		if (out_ref_valid_bits < 0)
+			return out_ref_valid_bits;
+
+		out_ref_rate = params_rate(fe_params);
+		out_ref_channels = params_channels(fe_params);
```

Comment:
can this happen? 

there's an earlier switch where the default case ends with an error.

https://github.com/thesofproject/linux/pull/4332/commits/5e87b60d1a5b280f6b5a45a1e413550a6b2f5dcd#diff-694481cfb73dd1682c248c54855215595dc96ca4ee65b1d3ef9cdc4a478158e0R1393

https://github.com/thesofproject/linux/pull/4332/commits/5e87b60d1a5b280f6b5a45a1e413550a6b2f5dcd#diff-694481cfb73dd1682c248c54855215595dc96ca4ee65b1d3ef9cdc4a478158e0R1511

Reply:
@plbossart the previoud switch case would have dealth with all of widgets that arent dai_in/dai_out, aif_in/aif_out and buffer. So when it comes to this switch, we only have cases to deal with. One where the ref is based on input audio format and the other where the ref is based on fe_params. So the 2 defaults you are looking at different.

Reply:
not really @ranj063, the flow is exactly the same in the two switch cases

````
switch (swidget->id) {
case snd_soc_dapm_aif_in: 
case snd_soc_dapm_aif_out:
case snd_soc_dapm_dai_in:
case snd_soc_dapm_dai_out:
case snd_soc_dapm_buffer:

default:
dev_err(sdev->dev, "unsupported type %d for copier %s",
		   swidget->id, swidget->widget->name);
return -EINVAL;
}

switch (swidget->id) {

case snd_soc_dapm_aif_in:
case snd_soc_dapm_dai_out:
case snd_soc_dapm_buffer:
case snd_soc_dapm_aif_out:
case snd_soc_dapm_dai_in:
default:
	out_ref_valid_bits = sof_ipc4_get_valid_bits(sdev, fe_params);
	if (out_ref_valid_bits < 0)
		return out_ref_valid_bits;

	out_ref_rate = params_rate(fe_params);
	out_ref_channels = params_channels(fe_params);
````

I will still assert the second case cannot happen.

Reply:
@plbossart you are right. sorry this was not my suggestion. Mine was different as in https://github.com/thesofproject/linux/pull/4332#discussion_r1180568418

Reply:
Hi, @ranj063 , to confirm, for 1st default case, instead of return -EINVAL, we now add the updated block right?